File: 144. Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    '''
    recrusion method
    '''
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        l = []
        self.helper(root, l)
        return l
        
    def helper(self, root, lis):
        if not root:
            return
        lis.append(root.val)
        self.helper(root.left, lis)
        self.helper(root.right, lis)

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        '''
        iterative method
        '''
        if not root:
            return []
        stack = [root]
        output = []
        while stack:
            node = stack.pop(-1)
            if node:
                output.append(node.val)
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        return output

File: 144. Binary Tree Preorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
/**
 * recursion method
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        helper(root, ans);
        return ans;
    }
    private void helper(TreeNode root, List<Integer> ans) {
        if (root == null) {
            return;
        }
        ans.add(root.val);
        helper(root.left, ans);
        helper(root.right, ans);
    }
}

class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        List<Integer> ans = new ArrayList<>();
        Deque<TreeNode> dq = new ArrayDeque<>();
        dq.add(root);
        while (!dq.isEmpty()) {
            TreeNode node = dq.pollLast();
            ans.add(node.val);
            if (node.right!=null) {
                dq.add(node.right);
            }
            if (node.left!=null) {
                dq.add(node.left);
            }
        }
        return ans;
    }
}